style/hudl_sql_style_guide.md

```diff
@@ -0,0 +1,339 @@
+---
+layout: default
+title: SQL Style Guide
+description: A guide to writing clean, clear, and consistent SQL.
+---
+
+# Purpose
+
+
+# Principles
+
+* We take a disciplined and practical approach to writing code.
+* We regularly check-in code to Github
+* We believe consistency in style is very important.
+* We demonstrate intent explicitly in code, via clear structure and comments where needed.
+
+# Rules
+
+## General
+
+* No tabs. 2 spaces per indent.
+* No trailing whitespace.
+* 80 character limit per line
+* Include links to hudl id constants or references to important links
+* All tables and views should adhere to all lowercase, words separated with underscores
+* Always use AS when creating aliases for columns, tables, subqueries
+* Always capitalize SQL keywords (e.g., `SELECT` or `AS`)
+* Variable names should be underscore separated:
+
+  __GOOD__:
+  `SELECT COUNT(*) AS backers_count`
+
+  __BAD__:
+  `SELECT COUNT(*) AS backersCount`
+
+* Comments should go near the top of your query, or at least near the closest `SELECT`
+* Try to only comment on things that aren't obvious about the query (e.g., why a particular ID is hardcoded, etc.)
+* Don't use single letter variable names be as descriptive as possible given the context:
+
+  __GOOD__:
+  `SELECT ksr.backings AS backings_with_creators`
+
+  __BAD__:
+  `SELECT ksr.backings AS b`
+
+* Use [Common Table Expressions](http://www.postgresql.org/docs/8.4/static/queries-with.html) (CTEs) early and often, and name them well.
+
+## `SELECT`
+
+Align all columns to the first column on their own line:
+
+```sql
+SELECT
+  projects.name,
+  users.email,
+  projects.country,
+  COUNT(backings.id) AS backings_count
+FROM ...
+```
+
+`SELECT` goes on its own line:
+
+```sql
+SELECT [DISTINCT]
+  name,
+  ...
+```
+
+Always rename aggregates and function-wrapped columns:
+
+```sql
+SELECT
+  name,
+  SUM(amount) AS sum_amount
+FROM ...
+```
+
+
+Long Window functions should be split across multiple lines: one for the `PARTITION`, `ORDER` and frame clauses, aligned to the `PARTITION` keyword. Partition keys should be one-per-line, aligned to the first, with aligned commas. Order (`ASC`, `DESC`) should always be explicit. All window functions should be aliased.
+
+```sql
+SUM(1) OVER (
+  PARTITION BY category_id,
+               year
+  ORDER BY pledged DESC
+  ROWS UNBOUNDED PRECEDING) AS category_year
+```
+
+## `FROM`
+
+Only one table should be in the `FROM`. Never use `FROM`-joins:
+
+__GOOD__:
+
+```sql
+SELECT
+  projects.name AS project_name,
+  COUNT(backings.id) AS backings_count
+FROM ksr.projects AS projects
+JOIN ksr.backings AS backings ON backings.project_id = projects.id
+...
+```
+
+__BAD__:
+
+```sql
+SELECT
+  projects.name AS project_name,
+  COUNT(backings.id) AS backings_count
+FROM ksr.projects AS projects, ksr.backings AS backings
+WHERE
+  backings.project_id = projects.id
```

Comment:
I think we agreed on 
```sql
WHERE backings.project_id = projects.id
```
and then for multiple clauses
```sql
WHERE backings.project_id = projects.id
  AND backings.other_id = projects.other_id
```

Reply:
I think that's right. This is in the "bad" examples section. Also, I just ported this over from kickstarter. Should we keep bad around or just keep good examples?

Reply:
Good point, I didn't notice it was in the bad section. I am confused why there is a bad example of `WHERE` in the `FROM` section though.